File: src/main/java/io/gitlab/intended/storagenetworks/inventory/CraftingTerminalInventory.java
package io.gitlab.intended.storagenetworks.inventory;

import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.util.Identifier;
import net.minecraft.util.PacketByteBuf;
import spinnery.common.BaseInventory;
import spinnery.widget.WInterface;
import spinnery.widget.WSlot;

public class CraftingTerminalInventory extends ModInventory {

    public static final int INVENTORY = 1;

    public CraftingTerminalInventory(int syncId, Identifier id, PlayerEntity player, PacketByteBuf buf) {
        super(syncId, id, player, buf);

        WInterface mainInterface = getInterface();
        BaseInventory inventory = new BaseInventory(27);

        getInventories().put(INVENTORY, inventory);
        mainInterface.createChild(WSlot::new);
        WSlot.addHeadlessArray(mainInterface, 0, INVENTORY, 9, 3);
        WSlot.addHeadlessPlayerInventory(mainInterface);
    }

}

File: src/main/kotlin/io/gitlab/intended/storagenetworks/inventory/CraftingTerminalInventory.kt
package io.gitlab.intended.storagenetworks.inventory

import net.minecraft.entity.player.PlayerEntity
import net.minecraft.text.Text
import spinnery.common.BaseInventory
import spinnery.widget.WSlot

class CraftingTerminalInventory(
    syncId: Int,
    player: PlayerEntity,
    name: Text
) : ModInventory(syncId, player, name) {

    companion object {
        const val INV = 1
    }

    init {
        val root = `interface`
        val inventory = BaseInventory(27)

        inventories[INV] = inventory

        root.createChild { WSlot() }
        WSlot.addHeadlessArray(root, 0, INV, 9, 3)
        WSlot.addHeadlessPlayerInventory(root)
    }

}